File: src/main/java/com/atsuishio/superbwarfare/data/gun/GunProp.java
package com.atsuishio.superbwarfare.data.gun;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.data.ObjectToList;
import com.atsuishio.superbwarfare.data.StringToObject;
import com.google.gson.annotations.SerializedName;
import net.minecraft.util.Mth;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Field;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

public class GunProp<T> {
    // 这b玩意必须放第一个，不然new的时候执行到props.add(this)会NPE（全恼
    private static final List<GunProp<?>> props = new ArrayList<>();

    public static final GunProp<Integer> MAX_DURABILITY = new GunProp<Integer>("MaxDurability", true)
            .withLimiter((data, v) -> Math.max(0, v));

    public static final GunProp<Double> RECOIL_X = new GunProp<>("RecoilX");
    public static final GunProp<Double> RECOIL_Y = new GunProp<>("RecoilY");
    public static final GunProp<Double> RECOIL = new GunProp<>("Recoil");

    public static final GunProp<Double> SPREAD = new GunProp<>("Spread");
    public static final GunProp<Double> DAMAGE = new GunProp<>("Damage");
    public static final GunProp<Double> HEADSHOT = new GunProp<>("Headshot");
    public static final GunProp<Double> VELOCITY = new GunProp<>("Velocity");
    public static final GunProp<Integer> MAGAZINE = new GunProp<>("Magazine");

    public static final GunProp<Double> MELEE_DAMAGE = new GunProp<>("MeleeDamage");
    public static final GunProp<Integer> MELEE_DURATION = new GunProp<Integer>("MeleeDuration")
            .withLimiter((data, v) -> Math.max(0, v));

    public static final GunProp<Integer> MELEE_DAMAGE_TIME = new GunProp<Integer>("MeleeDamageTime")
            .withLimiter((data, v) -> Math.min(data.get(MELEE_DURATION), v));

    public static final GunProp<ProjectileInfo> PROJECTILE = new GunProp<>("Projectile");
    public static final GunProp<Integer> PROJECTILE_AMOUNT = new GunProp<>("ProjectileAmount");
    public static final GunProp<Double> WEIGHT = new GunProp<>("Weight");
    public static final GunProp<FireMode> DEFAULT_FIRE_MODE = new GunProp<>("DefaultFireMode");
    public static final GunProp<Set<FireMode>> AVAILABLE_FIRE_MODES = new GunProp<>("AvailableFireModes");

    public static final GunProp<Double> DEFAULT_ZOOM = new GunProp<>("DefaultZoom");

    public static final GunProp<Integer> BURST_AMOUNT = new GunProp<>("BurstAmount");
    public static final GunProp<Double> BYPASSES_ARMOR = new GunProp<>("BypassesArmor");

    public static final GunProp<Integer> NORMAL_RELOAD_TIME = new GunProp<>("NormalReloadTime");
    public static final GunProp<Integer> EMPTY_RELOAD_TIME = new GunProp<>("EmptyReloadTime");
    public static final GunProp<Integer> BOLT_ACTION_TIME = new GunProp<>("BoltActionTime");
    public static final GunProp<Integer> PREPARE_TIME = new GunProp<>("PrepareTime");
    public static final GunProp<Integer> PREPARE_LOAD_TIME = new GunProp<>("PrepareLoadTime");
    public static final GunProp<Integer> PREPARE_AMMO_LOAD_TIME = new GunProp<>("PrepareAmmoLoadTime");
    public static final GunProp<Integer> PREPARE_EMPTY_TIME = new GunProp<>("PrepareEmptyTime");
    public static final GunProp<Integer> ITERATIVE_TIME = new GunProp<>("IterativeTime");
    public static final GunProp<Integer> ITERATIVE_AMMO_LOAD_TIME = new GunProp<>("IterativeAmmoLoadTime");
    public static final GunProp<Integer> ITERATIVE_LOAD_AMOUNT = new GunProp<>("IterativeLoadAmount");
    public static final GunProp<Integer> FINISH_TIME = new GunProp<>("FinishTime");

    public static final GunProp<Double> SOUND_RADIUS = new GunProp<>("SoundRadius");

    public static final GunProp<Integer> RPM = new GunProp<Integer>("RPM")
            .withLimiter((data, v) -> Mth.clamp(v, 1, 114514));

    public static final GunProp<Double> EXPLOSION_DAMAGE = new GunProp<>("ExplosionDamage");
    public static final GunProp<Double> EXPLOSION_RADIUS = new GunProp<>("ExplosionRadius");
    public static final GunProp<Double> GRAVITY = new GunProp<>("Gravity");

    public static final GunProp<Integer> SHOOT_DELAY = new GunProp<>("ShootDelay");
    public static final GunProp<Double> HEAT_PER_SHOOT = new GunProp<>("HeatPerShoot");

    public static final GunProp<List<String>> AVAILABLE_PERKS = new GunProp<>("AvailablePerks");

    private final String name;
    private final Field field;
    private final boolean readOnly;
    public GunPropModifyContext<T> limiter;

    private GunProp(String name) {
        this(name, false);
    }

    public Type getFieldType() {
        return this.field.getType();
    }

    private GunProp(String name, boolean readOnly) {
        this.name = name;
        this.readOnly = readOnly;

        try {
            this.field = Arrays.stream(DefaultGunData.class.getFields())
                    .filter(f -> {
                        var annotation = f.getAnnotation(SerializedName.class);
                        return annotation != null && annotation.value().equals(this.name);
                    })
                    .findFirst()
                    .orElseThrow();
            this.field.setAccessible(true);
        } catch (Exception exception) {
            Mod.LOGGER.error("Could not find field {} in DefaultGunData!", name);
            throw new RuntimeException(exception);
        }

        props.add(this);
    }

    private GunProp<T> withLimiter(GunPropModifyContext<T> limiter) {
        this.limiter = limiter;
        return this;
    }

    @SuppressWarnings("unchecked")
    public T getDefault(DefaultGunData data) {
        try {
            return (T) processValue(field.get(data));
        } catch (Exception exception) {
            Mod.LOGGER.error("Could not get field {} in DefaultGunData!", name);
            throw new RuntimeException(exception);
        }
    }

    private static Object processValue(Object value) {
        if (value instanceof ObjectToList<?> otl) {
            return otl.list.stream().map(GunProp::processValue).toList();
        } else if (value instanceof StringToObject<?> sto) {
            return processValue(sto.value);
        }
        return value;
    }

    public GunPropModifier<T> asModifier(GunData data) {
        return new GunPropModifier<>(data, this.getDefault(data.getDefault()), limiter, readOnly);
    }

    public static @Nullable GunProp<?> getByName(String name) {
        return props.stream().filter(p -> p.name.equals(name)).findFirst().orElse(null);
    }

    @FunctionalInterface
    public interface GunPropModifyContext<T> {
        T apply(@NotNull GunData data, @NotNull T value);
    }

    public static class GunPropModifier<T> {
        private final GunData data;
        private final T value;
        private final GunPropModifyContext<T> limiter;
        private final boolean readOnly;

        private final List<GunPropModifyContext<T>> modifiers = new ArrayList<>();

        private GunPropModifier(GunData data, T value, @Nullable GunPropModifyContext<T> limiter, boolean readOnly) {
            this.data = data;
            this.value = value;
            this.limiter = limiter;
            this.readOnly = readOnly;
        }

        public GunPropModifier<T> apply(@Nullable List<GunPropModifyContext<T>> modifiers) {
            if (modifiers == null || readOnly) return this;

            for (var modifier : modifiers) {
                apply(modifier);
            }
            return this;
        }

        public GunPropModifier<T> apply(@Nullable GunPropModifyContext<T> modifier) {
            if (modifier == null || readOnly) return this;

            modifiers.add(modifier);
            return this;
        }

        public GunPropModifier<T> override(@Nullable T value) {
            if (value == null || readOnly) return this;

            modifiers.add((data, v) -> value);
            return this;
        }

        public T compute() {
            if (readOnly) return value;

            var result = value;

            for (var modifier : modifiers) {
                result = modifier.apply(data, result);
            }

            if (limiter != null) {
                result = limiter.apply(data, result);
            }

            return result;
        }
    }

}

File: src/main/java/com/atsuishio/superbwarfare/data/gun/GunProp.kt
package com.atsuishio.superbwarfare.data.gun

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.data.ModColor
import com.atsuishio.superbwarfare.data.Prop
import com.atsuishio.superbwarfare.data.gun.GunData.Companion.getPerkPriority
import com.atsuishio.superbwarfare.init.ModPerks
import com.atsuishio.superbwarfare.item.gun.GunItem
import com.atsuishio.superbwarfare.perk.Perk
import net.minecraft.util.Mth
import net.minecraftforge.registries.RegistryManager
import kotlin.math.max
import kotlin.math.min
import kotlin.reflect.KMutableProperty1

@Suppress("UNUSED")
class GunProp<T, R> private constructor(
    prop: KMutableProperty1<DefaultGunData, T>,
    transform: (T) -> R,
    limiter: PropModifyContext<GunData, DefaultGunData, R>.() -> R = { value },
) :
    Prop<GunData, DefaultGunData, T, R, GunProp<T, R>>(prop, transform, limiter) {

    companion object {
        private fun <T> plainProp(
            prop: KMutableProperty1<DefaultGunData, T>,
            limiter: PropModifyContext<GunData, DefaultGunData, T>.() -> T = { value },
        ): GunProp<T, T> {
            return GunProp(prop, { it }, limiter)
        }

        @JvmField
        val MAX_DURABILITY = plainProp(DefaultGunData::maxDurability) { max(0, value) }

        @JvmField
        val DURABILITY_PER_SHOOT = plainProp(DefaultGunData::durabilityPerShoot) { max(0, value) }

        @JvmField
        val MAX_ENERGY = plainProp(DefaultGunData::maxEnergy) { max(0, value) }

        @JvmField
        val MAX_RECEIVE_ENERGY = plainProp(DefaultGunData::maxReceiveEnergy) {
            val maxEnergy = modifier.get(MAX_ENERGY)
            val value = Mth.clamp(value, -1, maxEnergy)
            if (value < 0) maxEnergy else value
        }

        @JvmField
        val MAX_EXTRACT_ENERGY = plainProp(DefaultGunData::maxExtractEnergy) {
            val maxEnergy = modifier.get(MAX_ENERGY)
            val value = Mth.clamp(value, -1, maxEnergy)
            if (value < 0) maxEnergy else value
        }

        @JvmField
        val RECOIL_X = plainProp(DefaultGunData::recoilX)

        @JvmField
        val RECOIL_Y = plainProp(DefaultGunData::recoilY)

        @JvmField
        val RECOIL = plainProp(DefaultGunData::recoil)

        @JvmField
        val RECOIL_TIME = plainProp(DefaultGunData::recoilTime)

        @JvmField
        val RECOIL_FORCE = plainProp(DefaultGunData::recoilForce)

        @JvmField
        val SHOOT_SHAKE = plainProp(DefaultGunData::shootShake)

        @JvmField
        val SPREAD = plainProp(DefaultGunData::spread)

        @JvmField
        val DAMAGE = plainProp(DefaultGunData::damage)

        @JvmField
        val HEADSHOT = plainProp(DefaultGunData::headshot)

        @JvmField
        val VELOCITY = plainProp(DefaultGunData::velocity)

        @JvmField
        val MELEE_DAMAGE = plainProp(DefaultGunData::meleeDamage)

        @JvmField
        val MELEE_DURATION = plainProp(DefaultGunData::meleeDuration) { max(1, value) }

        @JvmField
        val MELEE_RANGE = plainProp(DefaultGunData::meleeRange)

        @JvmField
        val MELEE_ANGLE = plainProp(DefaultGunData::meleeAngle) { value.coerceIn(1, 180) }

        @JvmField
        val ZOOM_SPREAD_RATE = plainProp(DefaultGunData::zoomSpreadRate) { value.coerceIn(0.0, 1.0) }

        @JvmField
        val SEEK_TIME = plainProp(DefaultGunData::seekTime)

        @JvmField
        val SEEK_ANGLE = plainProp(DefaultGunData::seekAngle)

        @JvmField
        val SEEK_RANGE = plainProp(DefaultGunData::seekRange)

        @JvmField
        val MIN_TARGET_HEIGHT = plainProp(DefaultGunData::minTargetHeight)

        @JvmField
        val MAX_TARGET_HEIGHT = plainProp(DefaultGunData::maxTargetHeight)

        @JvmField
        val RANGE = plainProp(DefaultGunData::range) { max(1, value) }

        @JvmField
        val MELEE_DAMAGE_TIME =
            plainProp(DefaultGunData::meleeDamageTime) { min(modifier.get(MELEE_DURATION) - 1, value) }

        @JvmField
        val PROJECTILE = GunProp(DefaultGunData::projectile, { it.value!! })

        @JvmField
        val AMMO_COST_PER_SHOOT = plainProp(DefaultGunData::ammoCostPerShoot) { max(0, value) }

        @JvmField
        val PROJECTILE_AMOUNT = plainProp(DefaultGunData::projectileAmount) { max(0, value) }

        @JvmField
        val WEIGHT = plainProp(DefaultGunData::weight) { max(1.0, value) }

        @JvmField
        val DEFAULT_FIRE_MODE = GunProp(DefaultGunData::defaultFireMode, { it.ifEmpty { FireMode.SEMI.name!! }!! })

        @JvmField
        val AVAILABLE_FIRE_MODES =
            GunProp(DefaultGunData::availableFireModes, { it?.list?.map { l -> l.value!! } ?: listOf() })

        @JvmField
        val MAGAZINE = plainProp(DefaultGunData::magazine) {
            if (modifier.get(PROJECTILE_AMOUNT) <= 0 && modifier.get(MELEE_DAMAGE) > 0) 0 else max(0, value)
        }

        @JvmField
        val RELOAD_TYPES = GunProp(DefaultGunData::reloadTypes, { it?.filterNotNull().orEmpty() })

        @JvmField
        val SEEK_TYPE = GunProp(DefaultGunData::seekType, { it ?: SeekType.NONE })

        @JvmField
        val GUN_TYPE = GunProp(DefaultGunData::gunType, { it ?: GunType.SPECIAL })

        // 注意Nullable
        @JvmField
        val AUTO_RELOAD = plainProp(DefaultGunData::autoReload)

        @JvmField
        val WITHDRAW_AMMO_WHEN_CHANGE_SLOT = plainProp(DefaultGunData::withdrawAmmoWhenChangeSlot)

        @JvmField
        val ZOOM_RELOAD = plainProp(DefaultGunData::zoomReload)

        @JvmField
        val CLEAR_HOLD_PROGRESS_AFTER_SHOOT = plainProp(DefaultGunData::clearHoldProgressAfterShoot)

        @JvmField
        val DEFAULT_ZOOM = plainProp(DefaultGunData::defaultZoom)

        @JvmField
        val BURST_AMOUNT = plainProp(DefaultGunData::burstAmount) { max(0, value) }

        @JvmField
        val BYPASSES_ARMOR = plainProp(DefaultGunData::bypassesArmor)

        @JvmField
        val AMMO_CONSUMER = GunProp(
            DefaultGunData::ammoConsumers,
            { it?.list?.map { l -> l.value!!.also { consumer -> consumer.init() } } ?: listOf() }
        )

        @JvmField
        val NORMAL_RELOAD_TIME = plainProp(DefaultGunData::normalReloadTime)

        @JvmField
        val EMPTY_RELOAD_TIME = plainProp(DefaultGunData::emptyReloadTime)

        @JvmField
        val BOLT_ACTION_TIME = plainProp(DefaultGunData::boltActionTime)

        @JvmField
        val PREPARE_TIME = plainProp(DefaultGunData::prepareTime)

        @JvmField
        val PREPARE_LOAD_TIME = plainProp(DefaultGunData::prepareLoadTime)

        @JvmField
        val PREPARE_AMMO_LOAD_TIME = plainProp(DefaultGunData::prepareAmmoLoadTime)

        @JvmField
        val PREPARE_EMPTY_TIME = plainProp(DefaultGunData::prepareEmptyTime)

        @JvmField
        val ITERATIVE_TIME = plainProp(DefaultGunData::iterativeTime)

        @JvmField
        val ITERATIVE_AMMO_LOAD_TIME = plainProp(DefaultGunData::iterativeAmmoLoadTime)

        @JvmField
        val ITERATIVE_LOAD_AMOUNT = plainProp(DefaultGunData::iterativeLoadAmount)

        @JvmField
        val FINISH_TIME = plainProp(DefaultGunData::finishTime)

        @JvmField
        val BURST_COOLDOWN = plainProp(DefaultGunData::burstCooldown)

        @JvmField
        val SOUND_RADIUS = plainProp(DefaultGunData::soundRadius)

        @JvmField
        val RPM = plainProp(DefaultGunData::rpm) { value.coerceIn(1, 114514) }

        @JvmField
        val EXPLOSION_DAMAGE = plainProp(DefaultGunData::explosionDamage)

        @JvmField
        val EXPLOSION_RADIUS = plainProp(DefaultGunData::explosionRadius)

        @JvmField
        val GRAVITY = plainProp(DefaultGunData::gravity)

        @JvmField
        val SHOOT_DELAY = plainProp(DefaultGunData::shootDelay)

        @JvmField
        val HEAT_PER_SHOOT = plainProp(DefaultGunData::heatPerShoot)

        @JvmField
        val NATURAL_COOLDOWN = plainProp(DefaultGunData::naturalCooldown)

        @JvmField
        val IN_WATER_COOLDOWN_RATE = plainProp(DefaultGunData::inWaterCooldownRate)

        @JvmField
        val IN_SNOW_COOLDOWN_RATE = plainProp(DefaultGunData::inSnowCooldownRate)

        @JvmField
        val IN_FIRE_COOLDOWN_RATE = plainProp(DefaultGunData::inFireCooldownRate)

        @JvmField
        val IN_LAVA_COOLDOWN_RATE = plainProp(DefaultGunData::inLavaCooldownRate)

        @JvmField
        val AVAILABLE_PERKS = GunProp(DefaultGunData::availablePerks, {
            val availablePerks = mutableListOf<Perk>()
            val perkNames = it?.list?.filterNotNull().orEmpty().ifEmpty { return@GunProp availablePerks }

            val sortedNames = perkNames.distinct().sortedWith { s1, s2 ->
                val p1 = getPerkPriority(s1)
                val p2 = getPerkPriority(s2)
                if (p1 != p2) {
                    return@sortedWith p1.compareTo(p2)
                } else {
                    return@sortedWith s1.compareTo(s2)
                }
            }

            val perks = RegistryManager.ACTIVE.getRegistry(ModPerks.PERK_KEY).getEntries()

            val perkValues = perks.mapNotNull { obj -> obj?.value }
            val perkKeys = perks.mapNotNull { perk -> perk?.key?.location().toString() }

            for (name in sortedNames) {
                if (name.startsWith("@")) {
                    when (name.substring(1)) {
                        "Ammo" -> Perk.Type.AMMO
                        "Functional" -> Perk.Type.FUNCTIONAL
                        "Damage" -> Perk.Type.DAMAGE
                        else -> null
                    }?.let { type ->
                        availablePerks.addAll(perkValues.filter { it.type == type })
                    }
                } else if (name.startsWith("!")) {
                    val n = name.substring(1)
                    val index = perkKeys.indexOf(n)
                    if (index != -1) {
                        availablePerks.remove(perkValues[index])
                    } else {
                        Mod.LOGGER.info("Perk {} not found", n)
                    }
                } else {
                    val index = perkKeys.indexOf(name)
                    if (index != -1) {
                        availablePerks.add(perkValues[index])
                    } else {
                        Mod.LOGGER.info("Perk {} not found", name)
                    }
                }
            }
            return@GunProp availablePerks.toList()
        })

        @JvmField
        val ICON = GunProp(DefaultGunData::icon, { it ?: GunItem.DEFAULT_ICON })

        @JvmField
        val CROSSHAIR = GunProp(DefaultGunData::crosshair, { it.ifEmpty { "@GunDefault" }!! })

        @JvmField
        val CROSSHAIR_ZOOMING = GunProp(DefaultGunData::crosshairZooming, { it.ifEmpty { "@Empty" }!! })

        @JvmField
        val CROSSHAIR_COLOR = GunProp(DefaultGunData::crosshairColor, { it ?: ModColor() })

        // 注意Nullable
        @JvmField
        val NAME = plainProp(DefaultGunData::name)

        @JvmField
        val SHOOT_POS = GunProp(DefaultGunData::shootPos, { it ?: ShootPos() })

        @JvmField
        val SEEK_WEAPON_INFO = plainProp(DefaultGunData::seekWeaponInfo)

        @JvmField
        val SOUND_INFO = GunProp(DefaultGunData::soundInfo, { it ?: SoundInfo() })

        @JvmField
        val SHOOT_ANIMATION_TIME = plainProp(DefaultGunData::shootAnimationTime)

        @JvmField
        val SPREAD_AMOUNT = plainProp(DefaultGunData::spreadAmount)

        @JvmField
        val SPREAD_ANGLE = plainProp(DefaultGunData::spreadAngle)

        @JvmField
        val ADD_SHOOTER_DELTA_MOVEMENT = plainProp(DefaultGunData::addShooterDeltaMovement)

        @JvmField
        val SHELL_TYPE = plainProp(DefaultGunData::shellType)
    }
}